### PR TITLE
fix(provider/kubernetes): add more checking for verifying runing instances.

### PR DESCRIPTION
**Pull request by chlung**
About this PR,

Add a new onDemand type for controller force refresh.

Thanks.
We prefer small, well tested pull requests.

Please refer to [Contributing to Spinnaker](https://spinnaker.io/community/contributing/).

When filling out a pull request, please consider the following:

* Follow the commit message conventions [found here](http://www.spinnaker.io/v1.0/docs/how-to-submit-a-patch).
* Provide a descriptive summary for your changes.
* If it fixes a bug or resolves a feature request, be sure to link to that issue.
* Add inline code comments to changes that might not be obvious.
* Squash your commits as you keep adding changes.
* Add a comment to @spinnaker/reviewers for review if your issue has been outstanding for more than 3 days.

Note that we are unlikely to accept pull requests that add features without prior discussion. The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy [hunk 1]
```diff
@@ -36,6 +36,7 @@ import static com.netflix.spinnaker.orca.ExecutionStatus.SUCCEEDED
 @Slf4j
 class ServerGroupCacheForceRefreshTask extends AbstractCloudProviderAwareTask implements RetryableTask {
   static final String REFRESH_TYPE = "ServerGroup"
+  static final String REFRESH_CONTROLLERS_TYPE = "ControllerSet"
```
Comment (lwander) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy]: Why can't the "controllerset" type accept refreshes for server groups instead of adding this new type here?

Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy]: We talked abou this.  I thought that you would like to break into two different request call.

Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy]: No maybe we miscommunicated, clouddriver should handle the single on demand update for any controller that can, which may be more than one.

**Comment by chlung** [comment]
@lwander updated with your feedback. Please review.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy [hunk 2]
```diff
@@ -43,6 +43,11 @@ class WaitForCapacityMatchTask extends AbstractInstancesCheckTask {
     if (!serverGroup.capacity || serverGroup.capacity.desired != instances.size()) {
       return false
     }
+
+    if (stage.context.get("capacity")?.get("desired") != instances.size()) {
```
Comment (lwander) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: If this key is missing, the stage will fail every time

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy [hunk 3]
```diff
@@ -158,7 +159,11 @@ class ServerGroupCacheForceRefreshTask extends AbstractCloudProviderAwareTask im
         }
 
         def forceCacheUpdate = pendingForceCacheUpdates.find {
-          (it.details as Map<String, String>).intersect(model) == model
+          if (it) {
```
Comment (lwander) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy]: Why is this needed?

Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy]: It is possible to send null to here because both agent is send data here here.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy (hunk 2 above)
Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: you mean context?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy [hunk 4]
```diff
@@ -43,6 +43,13 @@ class WaitForCapacityMatchTask extends AbstractInstancesCheckTask {
     if (!serverGroup.capacity || serverGroup.capacity.desired != instances.size()) {
       return false
     }
+
+    if (stage.context.get("capacity") && stage.context.get("capacity").get("desired")) {
```
Comment (lwander) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: Why doesn't the above check work? Shouldn't the server group have the capacity set?

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTaskSpec.groovy [hunk 5]
```diff
@@ -49,7 +49,7 @@ class WaitForCapacityMatchTaskSpec extends Specification {
       oort = Stub(OortService)
       oort.getServerGroup("test", "us-east-1", "kato-main-v000") >> { new Response('kato', 200, 'ok', [], new TypedString(mapper.writeValueAsString(serverGroup))) }
       task.oortService = oort
-      def context = [account: "test", "deploy.server.groups": ["us-east-1": ["kato-main-v000"]]]
+      def context = [account: "test", "deploy.server.groups": ["us-east-1": ["kato-main-v000"]], capacity: [desired: 3]]
```
Comment (lwander) [inline orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTaskSpec.groovy]: Are these changes still needed with the above null check?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy (hunk 4 above)
Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: This is possible my use case.  First, I have stage is set size to 0.  I think that it is timing during the resizing.  The above condition will satisfy and return true and it won't checking the size anymore and launch next stage.  So, some of the pod might not shutdown before remove the statefulset.

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTaskSpec.groovy (hunk 5 above)
Comment (chlung) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTaskSpec.groovy]: I will check

**Comment by chlung** [comment]
@lwander  I rollback unit test changes because null checking can resolve the exception.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy (hunk 4 above)
Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: @ajordens what are your thoughts on this change? I'm worried it'll have some weird consequences

Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: stage should have the desired instance.  return false will ask to recheck.  I can move the code the end before return true.  That might reduce some impact.  How do you think?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy (hunk 3 above)
Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy]: That seems like an error on the cloud provider agent, I can't think of a good reason why `pendingForceCacheUpdates` would contain a `null` value.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy (hunk 4 above)
Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: I'm not sure what we're checking here ... it looks like this change is relying on groovy truth and offering up some different behavior when `capacity.desired == 0`?

I don't believe we should short-circuit this task until the actual instance count matches desired.

Is that problem that you don't want to wait after you have resized to zero?

Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: My test case is for statefulset and daemon controllers which don't need disable serverGroup.
- create a pipeline
- create resize server group stage and set to 0 ( currently running 4 pods)
- create a destroy server group stage to delete controller.

First onDemand call,
(!serverGroup.capacity || serverGroup.capacity.desired != instances.size())  ==> true (4=4)

Now, we reach the following line.  serverGroup.disabled = false.  Now, it will return true and stop request any update and wait.
!serverGroup.disabled ?
      WaitForUpInstancesTask.allInstancesMatch(stage, serverGroup, instances, interestingHealthProviderNames) :
      true

Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: My goal is make sure to wait.  If I delete controller before pods terminated.  It will have orphan pods.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy (hunk 3 above)
Comment (chlung) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy]: I will take a look if it is possible filter in clouddriverside  if null.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy (hunk 4 above)
Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForCapacityMatchTask.groovy]: If you've resized to zero, `capacity.desired` would not equal `instances.count()` ... 0 != 4.

I must be missing something?

**Comment by chlung** [comment]
@lwander amd @ajordens - I found a bug in ControllerAgent(clouddriver side).  I will close this PR for now.  Thanks for your helps.
